src/finn/transformation/fpgadataflow/ip_cache.py

```diff
@@ -0,0 +1,642 @@
+    from qonnx.core.modelwrapper import ModelWrapper
+
+
+# UTILITY FUNCTIONS
+def _ndarray_to_bytes(tensor: Any) -> bytes:
+    cont = np.ascontiguousarray(tensor)
+    assert type(tensor) is np.ndarray
+    return cont.tobytes() + str(tensor.shape).encode("UTF-8")
+
+
+def _attribute_path_exists(name: str, op: HWCustomOp) -> bool:
+    """Check that the node attribute path exists.
+    If the node attribute cannot be loaded, return False."""  # noqa
+    try:
+        data = op.get_nodeattr(name)
+        if data is None or data == "":
+            return False
+        return Path(cast(str, data)).exists()
+    except Exception:
+        return False
+
+
+def _check_path_lengths_okay(
+    pc_name_max: int, pc_path_max: int, hashed_key: str, target_dir: Path
+) -> bool:
+    """Check if we follow the path length limits. If not return False, otherwise True."""
+    if len(hashed_key) > pc_name_max:
+        log.error(
+            f"Cannot cache an IP: The hash hex representation "
+            f"is too long to be allowed as a filename on your "
+            f"system (best effort detected limit: "
+            f"{pc_name_max}). Skipping caching."
+        )
+        return False
+    path_bytes = len(str(target_dir.absolute()).encode("UTF-8"))
+    if path_bytes > pc_path_max:
+        log.error(
+            f"Cannot cache an IP: the generated path length of "
+            f"the cache location is not allowed on your system! "
+            f"The best effort detected limit is: "
+            f"{pc_path_max} bytes, the path length is "
+            f"{path_bytes} bytes. Skipping caching."
+        )
+        return False
+    return True
+
+
+CACHE_IP_DEFINITIONS: dict[type, dict[str, list[str]]] = {}
+"""Contains all node attributes that a custom operator needs to be characterized.
+Filled by the cache_ip decorator. If the field "use" is defined, these attributes are
+used to hash the op.
+>>> CACHE_IP_DEFINITIONS[my_operator]["use"] = [...]
+
+However if "ignore" is used, every attribute _except_ those listed are used.
+>>> CACHE_IP_DEFINITIONS[my_operator]["ignore"] = [...]
+"""
+
+
+def cache_ip(attributes: list[str] | None = None) -> Callable[[type], type]:
+    """Decorate the given custom operator to be cacheable.
+
+    Args:
+        attributes: List of the key names of all node attributes needed to
+                    identify IP cores.
+    """
+    global CACHE_IP_DEFINITIONS
+
+    def wrapper(op_cls: type) -> type:
+        assert issubclass(
+            op_cls, HWCustomOp
+        ), f"Can only cache HWCustomOp instances, but {op_cls.__name__} is not a HWCustomOP!"
+        if op_cls not in CACHE_IP_DEFINITIONS.keys():
+            CACHE_IP_DEFINITIONS[op_cls] = {}
+        else:
+            # Already marked
+            return op_cls
+        if attributes is not None:
+            CACHE_IP_DEFINITIONS[op_cls]["use"] = attributes
+        else:
+            # List of fields that don't define the IP core itself,
```

Comment:
I think it should be safe to add the following attributes to this list of default ignored attributes. Most of them are optional (set to "" or [] by default) attributes of HWCustomOp, but they currently still show up in the key file.

backend
preferred_impl_style
exec_mode
rtlsim_trace
slr
mem_port
partition_id
device_id
inFIFODepths
outFIFODepths
output_hook
io_chrc_in
io_chrc_out
io_chrc_period
io_chrc_pads_in
io_chrc_pads_out